### PR TITLE
Make test source contents available on test object. Fixes gh-2

### PR DESCRIPTION
**Pull request by rwaldron**
This enables optimizations such as: 

```js
const stream = new TestStream(test262Dir, {
  includesDir,
  paths,
});

stream.on("data", test => {
  fs.readFile(path.join(test262Dir, test.file), "utf8", (error, contents) => {
    let ast = recast.parse(contents);
    // ...do stuff with ast, save new test source somewhere...
  });
});
```

To:  

```js
const stream = new TestStream(test262Dir, {
  includesDir,
  paths,
});

stream.on("data", test => {
  let ast = recast.parse(test.raw);
  // ...do stuff with ast, save new test source somewhere...
});
```

### CODE REVIEW

File: lib/compile.js [hunk 1]
```diff
@@ -28,8 +30,8 @@ module.exports = function (filePath, options, done) {
       test.insertionIndex = test.attrs.flags.raw ?
         -1 : test.contents.indexOf(test262StreamMarker);
       test.contents = test.contents.replace(test262StreamMarker, '');
-
       test.file = path.relative(options.test262Dir, test.file);
+      test.source = source;
```
Comment (leobalter) [inline lib/compile.js]: Naming is hard and I'm afraid to create a bikeshedding, but I wonder if this should be exposed as `test.raw` as the raw source. It helps to distinguish between contents and raw without much further documentation. 

`raw` also follows the flags rule used in Test262.

In any case, the current code works for me.

Comment (rwaldron) [reply lib/compile.js]: WFM (you can also push changes to my branch)

Comment (rwaldron) [reply lib/compile.js]: Actually, I can/should/will take care of this.

Comment (leobalter) [reply lib/compile.js]: Not bad, thanks!

**Comment by rwaldron** [comment]
@leobalter updated: source => raw

**Comment by jugglinmike** [comment]
Of the two scenarios implemented currently, one (the "strict" scenario) relies on a transformation of the test material. For tests that are expected to be suitable for both "strict" and "non-strict" scenarios, the usage you're describing will be inefficient (the same `raw` source will be emitted twice). For tests that describe "strict" scenarios only, it may be invalid--even for a consumer that is only interested in parsing (e.g. an `onlyStrict` test for an error like `var static;`).

I'm thinking about alternatives, and I feel we should probably avoid a solution that requires consumers to re-implement the scenario-generation logic. A better approach might involve a new configuration value--one that effects how each test is generated. Something like `skipIncludes` maybe.

What do you think?

**Comment by rwaldron** [comment]
> Something like skipIncludes maybe.

...is part of the solution, along with `skipScenarios`, which would be `false` by default, but when set to `true` causes only a single test object to be generated from a given source, with a `contents` property whose value is the raw test source. So one could write something like: 

```js
const stream = new TestStream(test262Dir, {
  includesDir,
  paths,
  skipIncludes: true, 
  skipScenarios: true, 
});

stream.on("data", test => {
  let ast = recast.parse(test.contents);
  // ...do stuff with ast, save new test source somewhere...
});
```

... Which would produce the data that I'm looking for.

**Comment by leobalter** [comment]
the API already tells me if it's strictOnly or otherwise. In this case, the script consuming it should know what to do with the raw code. 

Still, a skipIncludes behavior is slightly different and seems very interesting as well, where it produces the test _contents_ (not raw) without the includes.

Can we have both?

**Comment by leobalter** [comment]
my use case for raw is really about reporting. If something fails I can show only the raw contents of the failing test. In some comprehensive reporters this can be useful.

I also imagine this is of a lower priority, but it's already done here in this PR.

**Comment by rwaldron** [comment]
> where it produces the test contents (not raw) without the includes.

This is contrary to the use cases I have, I need just the raw test source (along with all of the other information that test262-stream provides in a test object). 

I have an alternate proposal to the alternate proposals: 

1. This PR moves forward, but behind a configuration property: `withSource: true | false` (or whatever, I don't care what we call it as long as the property doesn't include the word "include")
2. Implement `skipScenarios: true | false`, which would be `false` by default, but when set to `true` causes only a single test object to be generated from a given source, with a `contents` property whose value is the raw test source.

**Comment by leobalter** [comment]
1 and 2 works for me. It seems like skipScenarios might be interesting for my use case as well.

**Comment by rwaldron** [comment]
This is updated to my suggestion: 

> 1. This PR moves forward, but behind a configuration property: `withSource: true | false` (or whatever, I don't care what we call it as long as the property doesn't include the word "include")



```js
const stream = new TestStream(test262Dir, {
  includesDir,
  paths,
  withSource: true, 
});

stream.on("data", test => {
  let ast = recast.parse(test.source);
  // ...do stuff with ast, save new test source somewhere...
});
```

**Comment by rwaldron** [comment]
@leobalter I switched "raw" back to "source" because I don't want to conflate concepts with `flags: raw`.

**Comment by jugglinmike** [comment]
@rwaldron Can you say more about the use case for `{ withSource: true, skipScenarios: false }`?

**Comment by rwaldron** [comment]
> Can you say more about the use case for { withSource: true, skipScenarios: false }?

I have no interesting things to say about this, except that it produces the same result as my initial PR. Maybe someone wants that? Maybe we should throw an Error exception? "Don't do that"? I'm not really sure how to satisfy the design changes being requested of this work without losing the one thing that I really need: the untouched source code of the test.

**Comment by jugglinmike** [comment]
I'm reluctant to commit to an API which has an uneven surface area. Here are two other ideas that would avoid this. They might not work for you because I admittedly still don't understand your constraints very well.

One is that you use [the `klaw` module on npm](https://www.npmjs.com/package/klaw) directly to walk through the `test/` directory. This would certainly give you "the untouched source code of the test," but it would *not* give you access to the meta data. Not sure if that's a requirement for you, but I expect that it is.

Another option: we implement the `skipIncludes` option instead. That helps keep the configuration options independent so that users can't express invalid options in the first place. For your use case, you would not necessarily have "the untouched source code of the test" because many of the emitted tests will have an inserted `use strict` directive.

Depending on your needs, this may actually be an improvement over a solution based on `{withSource: true, skipScenarios: true}`. That's because you would *also* receive valid strict mode versions of the tests.

But it might be a deal breaker. I think you could use `skipIncludes` to mimic the behavior you are after:

```js
const stream = new Test262Stream(test262Dir, {
  includesDir,
  paths,
  skipIncludes: true
});

stream.on('data', (test) => {
  // The following two conditions are true for tests whose contents are a
  // modified version of the file on disk.
  if (test.attrs.flags.onlyStrict) {
    // This is a "strict mode" version of a test that does *not* have a
    // corresponding "non strict mode" version.
    //
    // This test is *only* valid after being transformed. You could either skip
    // it completely (i.e. `return` here), or you could use the transformed
    // version (i.e. remove this branch). I don't have enough information to
    // to say which of these is right for your use case.
  }
  if (!test.attrs.flags.onlyStrict && test.scenario === 'strict mode') {
    // This is a "strict mode" version of a test that also has a corresponding
    // "non strict mode" version.
    //
    // Skip it.
    return;
  }

  let ast = recast.parse(test.contents);
  // ...do stuff with ast, save new test source somewhere...
});
```

To be clear: this isn't completely equivalent to `{withSource: true, skipScenarios: true}` because it would not produce non-strict source for tests configured with `strictOnly`. That kind of output is invalid and potentially confusing, though, so I'd need to be convinced that it was appropriate for the stream to emit it.

**Comment by rwaldron** [comment]
I'm going to table this for now, and will continue to do it the hard way (which is fine).

**Pull request closed without merging**